lib/spack/spack/util/s3.py

```diff
@@ -4,83 +4,115 @@
-    # Gather dict of push URLS point to the value of the whole mirror
-    mirror_dict = {x.push_url: x for x in spack.mirror.MirrorCollection().values()}
-    # Ensure most specific URLs (longest) are presented first
-    mirror_url_keys = mirror_dict.keys()
-    mirror_url_keys = sorted(mirror_url_keys, key=len, reverse=True)
-    for mURL in mirror_url_keys:
-        # See if desired URL starts with the mirror's push URL
-        if rebuilt_path.startswith(mURL):
-            connection = mirror_dict[mURL].to_dict()[url_type]
-            break
-    return connection
 
+def get_s3_session(url, method="fetch"):
+    # import boto and friends as late as possible.  We don't want to require boto as a
+    # dependency unless the user actually wants to access S3 mirrors.
+    from boto3 import Session
+    from botocore import UNSIGNED
+    from botocore.client import Config
+    from botocore.exceptions import ClientError
 
-def _parse_s3_endpoint_url(endpoint_url):
-    if not urllib.parse.urlparse(endpoint_url, scheme="").scheme:
-        endpoint_url = "://".join(("https", endpoint_url))
-
-    return endpoint_url
+    # Circular dependency
+    from spack.mirror import MirrorCollection
 
+    global s3_client_cache
 
-def get_mirror_s3_connection_info(connection):
-    s3_connection = {}
-
-    s3_connection_is_dict = connection and isinstance(connection, dict)
-    if s3_connection_is_dict:
-        if connection.get("access_token"):
-            s3_connection["aws_session_token"] = connection["access_token"]
-        if connection.get("access_pair"):
-            s3_connection["aws_access_key_id"] = connection["access_pair"][0]
-            s3_connection["aws_secret_access_key"] = connection["access_pair"][1]
-        if connection.get("profile"):
-            s3_connection["profile_name"] = connection["profile"]
-
-    s3_client_args = {"use_ssl": spack.config.get("config:verify_ssl")}
-
-    endpoint_url = os.environ.get("S3_ENDPOINT_URL")
-    if endpoint_url:
-        s3_client_args["endpoint_url"] = _parse_s3_endpoint_url(endpoint_url)
-    elif s3_connection_is_dict and connection.get("endpoint_url"):
-        s3_client_args["endpoint_url"] = _parse_s3_endpoint_url(connection["endpoint_url"])
-
-    return (s3_connection, s3_client_args)
-
-
-def create_s3_session(url, connection={}):
+    # Get a (recycled) s3 session for a particular URL
     url = url_util.parse(url)
-    if url.scheme != "s3":
-        raise ValueError(
-            "Can not create S3 session from URL with scheme: {SCHEME}".format(SCHEME=url.scheme)
+
+    url_str = url_util.format(url)
+
+    def get_mirror_url(mirror):
+        return mirror.fetch_url if method == "fetch" else mirror.push_url
+
+    # Get all configured mirrors that could match.
+    all_mirrors = MirrorCollection()
+    mirrors = [
+        (name, mirror)
+        for name, mirror in all_mirrors.items()
+        if url_str.startswith(get_mirror_url(mirror))
```

Comment:
Maybe I get it.  This is where you've converted from the provide `url` to a mirror name, which you've possibly done in support of the future goal of using mirror names (rather than mirror urls) everywhere?

Reply:
Actually the old code was already like this, but maybe a comment got lost